### PR TITLE
Fixup test-datasets cache logic [KAT-3709]

### PR DESCRIPTION
**Pull request by e-mcginnis**
- retry more than once on failure
- introduce better logging
- cleanup in all fail scenerios

JIRA: KAT-3709

### CODE REVIEW

**Comment by e-mcginnis** [comment]
Immediately hit a networking flake in apt, how ironic

File: python/katana/example_data.py [hunk 1]
```diff
@@ -105,17 +106,25 @@ def _get_dataset(rel_path):
...
-        raise ValueError(f"Dataset does not exist at {path}")
+    while not path.exists() or count < max_attempts:
+        try:
+            # lame backoff, costs nothing on first attempt
+            time.sleep(count)
+            datasets_path = _get_test_datasets_directory(invalidate_cache=invalidate_cache)
+            path = datasets_path / rel_path
+            return path
+        except ValueError as e:
+            print(f"Observed {e} while getting test datasets, ignoring and trying again")
```
Comment (arthurp) [inline python/katana/example_data.py]: Use `logging`. https://docs.python.org/3/howto/logging.html#logging-basic-tutorial
Please use the `logger = logging.getLogger(__name__)` convention.

File: python/katana/example_data.py [hunk 2]
```diff
@@ -150,6 +159,7 @@ def find_path() -> Path:
         found_path = None
         # If SRC_DIR or KATANA_SOURCE_DIR environment is set, just use it
         env_path = []
+        print(f"os.environ = {os.environ}")
```
Comment (arthurp) [inline python/katana/example_data.py]: This shouldn't even be logged like this. Just remove this. If you want to log specific environment variables. That's fine.

File: python/katana/example_data.py [hunk 3]
```diff
@@ -173,13 +183,13 @@ def download_datasets(invalidate_cache=False) -> Path:
...
-                    shutil.rmtree(cache_path)
-                except OSError:
-                    cache_path.unlink()
 
+        # failed to validate, cleanup just in case
+        if cache_path.exists():
+            try:
+                shutil.rmtree(cache_path)
+            except OSError:
+                cache_path.unlink(missing_ok=True)
```
Comment (arthurp) [inline python/katana/example_data.py]: Do you want try-finally here?

**Review by arthurp — COMMENTED** [review]
Please switch to logging before merging this. We don't need print spam in our tests. Looks OK though other than that.

File: python/katana/example_data.py (hunk 1 above)
Comment (pyrito) [inline python/katana/example_data.py]: Try not to use `print`. I think Arthur wanted us to use the `logging` module instead and add a TODO for Arthur there as well.

File: python/katana/example_data.py (hunk 2 above)
Comment (pyrito) [inline python/katana/example_data.py]: Debugging print?

File: python/katana/example_data.py [hunk 4]
```diff
@@ -105,17 +109,25 @@ def _get_dataset(rel_path):
...
-        raise ValueError(f"Dataset does not exist at {path}")
+    while not path.exists() and count < max_attempts:
+        try:
+            # lame backoff, costs nothing on first attempt
+            time.sleep(count)
+            datasets_path = _get_test_datasets_directory(invalidate_cache=invalidate_cache)
+            path = datasets_path / rel_path
+            return path
+        except ValueError as e:
+            logger.info(f"Observed {e} while getting test datasets, ignoring and trying again")
```
Comment (arthurp) [inline python/katana/example_data.py:129]: NIT: You can include the exception in the log message with some argument which I cannot remember. Not critical though.

**Review by arthurp — APPROVED** [review]
LGTM.

**Pull request merged**